### DISCUSSION TITLE
Capturing the border pixels of an image

### DISCUSSION THREAD

**Discussion by keto33** [ImageMagick/ImageMagick#3175] [Help] [answered]
I want to check if a photo has a white background. I calculate the percentage of white colour by

`magick photo.jpg -threshold 99% -format "%[fx:100*image.mean]" info:`

This gives a good approximation, but to increase the accuracy, I want to calculate the percentage of white pixels at the border of the image (the top and bottom rows, the left and right columns). 

For example, 

![693-08127335en_Masterfile](https://user-images.githubusercontent.com/66027046/106548494-00f27c00-6507-11eb-8019-e65b285e3b2d.jpg)

I can crop four regions of 1px at North, South, West, and East; but it should be overkill. Is there a way to catch all pixels at the border of an image?

**Comment by fmw42** [comment]
Either of the following work to get the average intensity around the sides of the image. One turns on an opaque alpha channel. Then draws a black rectangle 1 pixel smaller all around inset 1 pixel in the alpha channel. Then scale the image to one pixels (it ignores transparent pixels) and then read out the pixel value.

```
magick rose: -alpha set -channel A -fill black -draw "rectangle 1,1 %[fx:w-2],%[fx:h-2]" +channel -scale 1x1 -alpha off -format "%[fx:intensity]" info:
0.0289176

```

```
magick rose: -alpha set -channel A -fill black -draw "rectangle 1,1 %[fx:w-2],%[fx:h-2]" +channel -colorspace gray -scale 1x1 -alpha off -format "%[fx:u.p]" info:
0.0289176

```

**Comment by GeeMack** [comment, marked as answer]
This command will isolate the single columns of pixels from the left and right, and the single rows of pixels from top and bottom, append them into a single image, then calculate the mean of that image...

    magick photo.jpg -roll +1+1 ( -clone 0 -extent 2x0 -rotate 90 ) ^
    ( -clone 0 -extent 0x2 ) -delete 0 +append -format "%[fx:mean*100]" info:

The "-roll" operation moves the far right column of pixels to the left edge, and the bottom row is moved to the top. Now the top two rows and the leftmost two columns contain all the pixels that were in the single pixel perimeter of the input image.

Next that result is cloned inside the first parentheses and the leftmost two columns are cropped and retained with "-extent" and rotated 90 degrees. Then in another set of parentheses the image is cloned and the top two rows are cropped and retained with "-extent".

After the parentheses the original input is deleted, leaving those two cropped images, one containing all the left and right column pixels, and the other containing all the top and bottom row pixels. Use "+append" to assemble those, then use your FX expression to calculate the mean of the four original edges from that.

If you're on a *nix system you'll need to escape the parentheses with backslashes "\\(...\\)" and change the continued-line caret "^" to a backslash "\\".
